### PR TITLE
Scan photos of specific directory

### PR DESCRIPTION
**Pull request by tetebueno**
I've been getting memory exhaustion when scanning all photos collection on an initial load; trying to address #525 I made these changes and was able to scan each year/month directory of photos succesfully like this (the example is for the folder with photos of july '84):

```console
$ docker exec -t -u www-data container-name php occ maps:scan-photos -vvv username Photos/1984/06
```

I didn't like the `if`s all over the implementation, but it's something like a _make it work_ approach. If I can get comments about it, we can make it better.

Cheers.

### CODE REVIEW

**Comment by tacruc** [comment]
Thanks for your contribution. Can you pass the path as optional argument?
Do you need help to fix the DCO?

**Comment by tetebueno** [comment]
You are very welcome.

Done with DCO, thanks.

Is there anything else besides [this](https://github.com/nextcloud/maps/pull/1231/files#diff-1e9219122ae9190e4bb16b6f373272412fdbb5d599ba78bd7480c5672ae45aa8R62) so the path parameter is optional? I believe it already is since help command displays it like this: `... [<user_id> [<path>]]`.

File: lib/Command/RescanPhotos.php [hunk 1]
```diff
@@ -77,18 +82,27 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
         $userId = $input->getArgument('user_id');
+        $pathToScan = $input->getArgument('path');
 		$inBackground = !($input->getOption('now') ?? true);
 		if ($inBackground) {
 			echo "Extracting coordinates from photo is performed in a BackgroundJob \n";
 		}
         if ($userId === null) {
             $this->userManager->callForSeenUsers(function (IUser $user) use ($inBackground) {
-                $this->rescanUserPhotos($user->getUID(), $inBackground);
+                if ($pathToScan === null) {
```
Comment (tacruc) [inline lib/Command/RescanPhotos.php:92]:
```suggestion
                $this->rescanUserPhotos($user->getUID(), $inBackground, pathToScan);
```

File: lib/Command/RescanPhotos.php [hunk 2]
```diff
@@ -77,18 +82,27 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
             });
         } else {
             $user = $this->userManager->get($userId);
             if ($user !== null) {
-                $this->rescanUserPhotos($userId, $inBackground);
+                if ($pathToScan === null) {
+                    $this->rescanUserPhotos($userId, $inBackground);
+                } else {
+                    $this->scanUserPhotos($userId, $pathToScan, $inBackground);
+                }
```
Comment (tacruc) [inline lib/Command/RescanPhotos.php:105]:
```suggestion
                $this->rescanUserPhotos($userId, $inBackground, pathToScan);
```

**Review by tacruc — CHANGES_REQUESTED** [review]
Hi @tetebueno,
sorry for the late reply and beeing a little unprecise.
I mean in stead of adding the functions
RescanPhotos->scanUserPhotos
and
PhotofilesService->rescanPath

Just provide an optional parameter to
```suggestion
       private function rescanUserPhotos(string $userId, bool $inBackground=true, sting? $pathToScan===null) {
```
and PhotofilesService
`public function rescan($userId, $inBackground=true, $pathToScan=null) {`


and then make an if in the Photofilesservice->rescan

```
$userFolder = $this->root->getUserFolder($userId)
if ($pathToScan === null) {
          $folder = $userFolder
} else {
          $folder = $userFolder->get($pathToScan);
}
$photos = $this->gatherPhotoFiles($folder, true);
```

Don't blindly copy my code. It is just typed in here to outline my thought and not tested.

**Comment by tacruc** [comment]
@tetebueno Do you think you will make the changes befor the NC29 release. Then I would wait with https://github.com/nextcloud/maps/pull/1244 to include this.

I'm sorry that it is on such short notice after not reviewing for a month. Just have to decide whether to wait or not.

**Comment by marlemion** [comment]
I just wanted to bring this back to mind. In my case it is tedious to update ~100000 Pictures just to get some recently geotagged pictures into Maps....

**Comment by tetebueno** [comment]
Apologies to everyone. Totally forgot about this.

@tacruc, I'm not completely sure what does this line do in `RescanPhotos+rescanUserPhotos()`, but I tried to maintain the logic intended in the first PR version:
```php
$this->config->setUserValue($userId, 'maps', 'installScanDone', 'yes');
```

File: lib/Command/RescanPhotos.php (hunk 1 above)
Comment (tetebueno) [reply lib/Command/RescanPhotos.php]: Already applied in last commit.

**Comment by tetebueno** [comment]
Forgot DCO again and it seems I went too far with it double-signing-off everything.

**Comment by tetebueno** [comment]
Ok, my bad, tried merging with Github's visual tool and it was an indentation mess ~~challenge~~. Will take care of this as soon as I can.

**Comment by tetebueno** [comment]
Finally got some time to pull this to a local environment and merge and fix formatting issues. Hope this is it.

**Review by tacruc — APPROVED** [review]

**Comment by tacruc** [comment]
@tetebueno  Thanks alot for your contribution. I will see if I find time the next two weeks to publish them. But I cannot promise it.

**Comment by tetebueno** [comment]
Gaaa, just got a chance to run `composer run cs:fix` and pushed the fixed code. Thanks anyway. I'm available for any issues regarding this.

**Pull request merged**